x/stakeibc/keeper/msg_server.go

```diff
@@ -1103,3 +1106,66 @@ func (k msgServer) ResumeHostZone(goCtx context.Context, msg *types.MsgResumeHos
 
 	return &types.MsgResumeHostZoneResponse{}, nil
 }
+
+// Submits an ICA tx to either grant or revoke authz permisssions to an address
+// to execute trades on behalf of the trade ICA
+func (k msgServer) ToggleTradeController(
+	goCtx context.Context,
+	msg *types.MsgToggleTradeController,
+) (*types.MsgToggleTradeControllerResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Fetch the trade ICA which will be the granter
+	tradeRoute, found := k.GetTradeRouteFromTradeChainId(ctx, msg.ChainId)
+	if !found {
+		return nil, types.ErrTradeRouteNotFound.Wrapf("trade route not found for chain ID %s", msg.ChainId)
+	}
+
+	swapMsgTypeUrl := "/" + proto.MessageName(&types.SwapAmountInRoute{})
```

Comment:
could you explain how this works?

Reply:
It get's the type URL of a message type (e.g. `/osmosis.gamm.v1beta1.MsgSwapExactAmountIn`)

Reply:
ok I understand now. do you think we should unit test? assembling the msg feels kinda complicated

Reply:
did you see the [unit tests](https://github.com/Stride-Labs/stride/pull/1142/commits/33b8ca1acff2193b84f0d5f57ec312b36a50264c#diff-1629187c35b24ec74868f7e895ab32337777eaf56723e246d834f27e31e7d75eR41) that are already in there?

Reply:
i missed those - they look good ty!